### PR TITLE
stdenv: Disallow output changes in installCheck

### LINKED ISSUE #120053: Outputs seem to be writable during installCheckPhase
**Describe the bug**

Badly written packages can modify the outputs during `installCheckPhase`. Packages won't be writable after the derivation completes, so the install check is not representative

**To Reproduce**

1. `nix-build --expr 'with import ./. {}; stdenv.mkDerivation { name = "x"; dontUnpack = true; doInstallCheck = true; installPhase = "mkdir $out"; installCheckPhase = "echo a >$out/a"; }'`
2. observe that it succeeds
3. `cat result/a`
4. observe that the package has a file that _probably_ shouldn't exist

**Expected behavior**

The build fails because the program tried to write to an output.

In `nixpkgs` this could be achieved with `chmod a-w`, except for stubborn packages that try really hard to write to their installation. A paranoid implementation could hash the files and check afterwards. Finally, Nix itself could provide a feature to make an output read-only when the builder requests it. This may not work on all platforms though.

**Metadata**

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - stdenv
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Spot installation script problems and avoid tests polluting the installation.

This will indicate errors in the install script fixed up by the
installed software, or software or configuration that relies on
a writable installation directory.

NOTE: this only works for derivations which use the installCheck phase as intended. Some language integrations, like `pythonPackages`, need to call the new functions on their own.

Closes #120053

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more ~NixOS~ test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/test/default.nix [hunk 1]
```diff
@@ -21,6 +21,7 @@ with pkgs;
   cc-wrapper-clang-9 = callPackage ./cc-wrapper { stdenv = llvmPackages_9.stdenv; };
   cc-wrapper-libcxx-9 = callPackage ./cc-wrapper { stdenv = llvmPackages_9.libcxxStdenv; };
   stdenv-inputs = callPackage ./stdenv-inputs { };
+  stdenv = callPackage ./stdenv { };
```
Comment (r-burns) [inline pkgs/test/default.nix:24]: No such file?

Comment (roberth) [reply pkgs/test/default.nix:24]: Oops. Added.

**Comment by symphorien** [comment]
Maybe we can chmod -w the outputs before installCheckPhase as well. What do you think ?

**Comment by roberth** [comment]
@symphorien I think we should, but it's not as simple as that because we'll have to restore the permissions, as `distPhase` needs to write to outputs again. I'd like to keep it simple for now.

**Comment by asymmetric** [comment]
@roberth how can i test this? I tried:

```
❯ nix build .#tests
error: flake output attribute 'legacyPackages.x86_64-linux.tests' is not a derivation
```

There are some ofborg eval failures, any idea why that is?

Sorry that I can't provide any more direct help at the moment.

**Comment by roberth** [comment]
> nix build .#tests

Tests contains a nested attrset of derivations. Try with `nix-build -A tests`, which traverses those attrs and doesn't copy all of your local nixpkgs into the store.

> ofborg eval failures

I don't think it's technically an eval failure, but rather that the eval status combines the other statuses. The `lib-tests` job did fail because it timed out rebuilding the world.

You could "just" build a package, rebuilding whatever needs to rebuild. A lot. If the build works, the pr works, but I don't know if this will impact many subtly broken tests. Maybe this pr should run separately on hydra first?

**Comment by symphorien** [comment]
This needs doc for allowWriteInInstallCheck and I think this is too late for 22.05 given the feature freeze #167025

**Comment by timokau** [comment]
Hi!

The marvin-mk2 bot is now [discontinued](https://github.com/timokau/marvin-mk2/issues/34#issuecomment-1100656280). I have removed the relevant tags from this PR. If you still need someone to look at it, one option would be to ask in [this discourse thread](https://discourse.nixos.org/t/prs-ready-for-review/3032).

I am posting this notice to all open PRs with the `marvin` tag. Please understand that following all of these discussions would take too much time, so I will unsubscribe from this PR unless I have already been involved in it before this message.

**Comment by roberth** [comment]
I did the thing.

**Pull request closed without merging**